frontend/lib/resolution/scope-queries.cpp

```diff
@@ -840,11 +843,33 @@ bool LookupHelper::doLookupInScope(const Scope* scope,
     traceCurPathSize = traceCurPath->size();
   }
 
-  auto pair = checkedScopes.insert(std::make_pair(name, scope));
-  if (pair.second == false) {
-    // scope has already been visited by this function,
-    // so don't try it again.
-    return false;
+  // update the checkedScopes map and return early if there is nothing to do.
+  auto p = checkedScopes.insert(std::make_pair(CheckedScope(name, scope),
+                                curFilter));
+  if (p.second == false) {
+    // the insert did not succeed: there was already something in the map.
+    // decide what to do about it.
+    IdAndVis::SymbolTypeFlags foundFilter = p.first->second;
+    if ((curFilter & foundFilter) == foundFilter) {
+      // if the flags we found are equal to foundFilter,
+      // or if curFilter is a superset of foundFilter
+      // (which, because these are filters, means that foundFilter is
+      //  less restricted / more general),
+      // there is no need to visit this scope further.
+      return false;
+    }
+
+    // otherwise, compute the new filter to use
+    IdAndVis::SymbolTypeFlags onlyInFound = 0;
+    onlyInFound = foundFilter & ~curFilter;
```

Comment:
It is not possible for `foundFilter={PRIVATE}`. I will adjust the example to use `foundFilter={PUBLIC}` to see if the logic is still right.

Given `foundFilter = {PUBLIC}` and `curFilter = {METHODS_OR_FIELDS}`, that means:
 * `foundFilter = {PUBLIC}` indicates that we already searched with the filter "Public fields only" (Note: regarding my "it is not possible" comment above there is not a situation where we *only* search "private fields only"; we can search for "private fields only" only if we are searching a scope that we previously searched for "public fields only" but we are finding a 2nd way to access it that can access private fields).
 * `curFilter = {METHODS_OR_FIELDS}` indicates that it is requested that we only consider methods and fields in this search.

What I want to compute in this example:
 * new setting for `curFilter`: I want it to search in the scope for private methods/fields. So that is new `curFilter'={PRIVATE,METHODS_OR_FIELDS}`.
   * you are correct that it is currently computing `onlyInFound={PUBLIC}` and new `curFilter'={PRIVATE}`.
 * it will compute `combinedFilter={}`; however this is arguably an approximation. We will have searched `{PUBLIC}` U `{PRIVATE,METHODS_OR_FIELDS}`, but a future search for `{PRIVATE}` would need to be able to find private non-methods. But, I think this kind of situation cannot happen in our scope resolver. It is a bit of a puzzle if we can add asserts to check for it or if we need to handle it in a different way.

Reply:
Yes, I was debating about `combinedFilter` with the same logic you presented, but managed to convince myself it was fine 🤦